tests/driver_at86rf2xx/main.c

```diff
@@ -20,31 +20,102 @@
 
 #include <stdio.h>
 
+#include "net/netdev2.h"
 #include "shell.h"
 #include "shell_commands.h"
-#include "net/gnrc/pktdump.h"
-#include "net/gnrc.h"
+#include "thread.h"
+#include "xtimer.h"
 
-/**
- * @brief   Maybe you are a golfer?!
- */
-int main(void)
+#include "common.h"
+
+#define _STACKSIZE      (THREAD_STACKSIZE_DEFAULT + THREAD_EXTRA_STACKSIZE_PRINTF)
+#define MSG_TYPE_ISR    (0x3456)
+
+static char stack[_STACKSIZE];
+static kernel_pid_t _recv_pid;
+
+at86rf2xx_t devs[AT86RF2XX_NUM];
+
+static const shell_command_t shell_commands[] = {
+    { "ifconfig", "Configure netdev2", ifconfig },
+    { "txtsnd", "Send IEEE 802.15.4 packet", txtsnd },
+    { NULL, NULL, NULL }
+};
+
+static void _event_cb(netdev2_t *dev, netdev2_event_t event, void *data)
 {
-    gnrc_netreg_entry_t dump;
+    (void) data;
+
+    if (event == NETDEV2_EVENT_ISR) {
+        msg_t msg;
+
+        msg.type = MSG_TYPE_ISR;
+        msg.content.ptr = (void *) dev;
+
+        if (msg_send(&msg, _recv_pid) <= 0) {
+            puts("gnrc_netdev2: possibly lost interrupt.");
+        }
+    }
+    else {
+        switch (event) {
+            case NETDEV2_EVENT_RX_COMPLETE:
+            {
+                recv(dev);
 
+                break;
+            }
+            default:
+                puts("Unexpected event received");
+                break;
+        }
+    }
+}
+
+void *_recv_thread(void *arg)
+{
+    while (1) {
+        msg_t msg;
+        msg_receive(&msg);
+        if (msg.type == MSG_TYPE_ISR) {
+            netdev2_t *dev = (netdev2_t *)msg.content.ptr;
+            dev->driver->isr(dev);
+        }
+        else {
+            puts("unexpected message type");
+        }
+    }
+}
+
+int main(void)
+{
     puts("AT86RF2xx device driver test");
+    xtimer_init();
+
+    for (unsigned i = 0; i < AT86RF2XX_NUM; i++) {
+        const at86rf2xx_params_t *p = &at86rf2xx_params[i];
+        netdev2_t *dev = (netdev2_t *)(&devs[i]);
```

Comment:
But `devs[i].netdev` isn't `netdev2_t` either, but `netdev2_ieee802154_t` so the cast would still be necessary and it would still be as hacky as this solution. That's the problem with the "super class as first member" paradigm in general: it simplifies maintainence of the inheritance, but the usage code always looks either hacky or ugly...